### ISSUE TITLE
How have you replaced colours on models so they are correct?

### ISSUE THREAD

**Issue by mattlewer** [Qodat/qodat#19]
More of a question than an issue.

I've been working on a personal project using models extracted from the cache. I've exported OBJ files and MTL files for all models and also exported the indexing files. Each asset in the indexing has the following fields (or similar), so that models can be reused but recoloured for different assets:
```
# Example from item Adamant Scimitar
 "colorFind": [
      61
  ],
  "colorReplace": [
      21662
  ],
```
I've figured out that the numbers are for HSL thanks to this useful tool & found the code in RuneLite that converts them:
https://rebecca-davies.github.io/color-picker/
https://github.com/runelite/runelite/blob/master/runelite-api/src/main/java/net/runelite/api/JagexColor.java

But still have a couple questions I hope you can solve

1. Sometimes a colorToFind or colorToReplace value is negative, why?
2.  Does colorFind[0] get replace by colorToReplace[0]? (asusme so)
3. If I run a colourToReplace number through Jagex's color converter, the result is close, but not the same as the colour in the MTL file? 

Example of question 3:

Color to replace:  
`61`
Original colour this refers to in MTL file:
`0.5411764705882353 0.48627450980392156 0.4823529411764706`
Result of converting 61:
`0.5420057871847349 0.48427139113337286 0.48426382451768296`

It's painfully close but not identical, if this something you've encountered?

Massively appreciate any help!

**Comment by dosier** [comment]
Heya sorry for the late reply!

> Sometimes a colorToFind or colorToReplace value is negative, why?

RS represents colours as shorts not as integers (to save 2 bytes).
Java by default uses signed types, so the range of possible vales is `-32768..32768`.
But you could represent the colour as an unsigned type as well, then you add`32768` to the value, making the new range `0..65.536`.

> Does colorFind[0] get replace by colorToReplace[0]? (asusme so)
Yes, the element at `colorToFind[index]` is replaced with `colorToReplace[index]`.

> If I run a colourToReplace number through Jagex's color converter, the result is close, but not the same as the colour in the MTL file?

Yes, jagex uses a custom colour space, they have a limited palette, details on how to generate this can be foundn the client code in one of the Rasterizer classes. But the values do not perfectly map onto each other, so some might look weird in a more standard color space.

This is something I hace yet to solve properly in qodat as well.

**Comment by mattlewer** [comment]
Thank you for your reply! Not many places to find help for thing like this, really appreciate it.

Sorry for not closing, I think I've managed to solve most of my issues by using euclidian distance to compare my output to the numbers in the MTL file for each object, I also needed to adjust my converter slightly :) Working for the vast majority of models, but still a few issues to sort... Onto figuring out animations!

If you're curious as to what I've made, let me know what you think (shameless plug I know)
https://gielinor3d.web.app/
